### PR TITLE
Discover nodes run by systemd and tail journald logs

### PR DESCRIPTION
**Pull request by iogakos**

### CODE REVIEW

File: internal/pkg/watch/watch.go [hunk 1]
```diff
@@ -124,3 +129,57 @@ func (w *Watch) Emit(message interface{}) {
...
+		// for the same buffer
+		break
+	}
+}
+
+func (w *Watch) emitLogStatusEvent(name string) {
+	ctx := map[string]interface{}{
+		model.NodeIDKey:      global.BlockchainNode.NodeID(),
+		model.NodeTypeKey:    global.BlockchainNode.NodeRole(),
+		model.NodeVersionKey: global.BlockchainNode.NodeVersion(),
```
Comment (tomasger) [inline internal/pkg/watch/watch.go:174]: There are cases when containers are undiscovered and these fields are empty. Perhaps we don't want to have empty context values arrive to our platform?

File: internal/pkg/watch/journal_logs.go [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+
+			log := w.Log.With("node_type", nodeType)
+			if global.BlockchainNode.LogWatchEnabled() {
+				if err := DefaultWatchRegistry.RegisterAndStart(w, subscriptions...); err != nil {
+					log.Errorw("failed to register journal log watcher", zap.Error(err))
+					return
+				}
+				log.Info("journal log watch started")
+			} else {
+				log.Info("docker log watch disabled - node type does not require logs to be watched")
```
Comment (tomasger) [inline internal/pkg/watch/journal_logs.go:250]:
```suggestion
				log.Info("journal log watch disabled - node type does not require logs to be watched")
```

File: protobind/node.go.template [hunk 3]
```diff
@@ -21,6 +21,21 @@ import (
...
+    // DefaultDiscoveryHintsSystemd default regular expression to detect Flow nodes run by systemd
+	DefaultDiscoveryHintsSystemd = "flow-(access|collection|verification|execution|consensus).service"
+
+    // DefaultDiscoveryHintsDocker default regular expression to detect Flow nodes run by docker
+	DefaultDiscoveryHintsDocker = "flow-go"
+)
+{{ else if eq .Blockchain "solana" }}
+var (
+	DefaultDiscoveryHintsSystemd = "solana-validator.service"
+	DefaultDiscoveryHintsDocker = "solana-validator"
```
Comment (tomasger) [inline protobind/node.go.template:35]: Should add comments here to avoid future linter errors in CI.

File: configs/agent.yml [hunk 4]
```diff
@@ -116,3 +116,11 @@ runtime:
...
     - type: prometheus.uname
     - type: prometheus.vmstat
+
+discovery:
+  hints:
+    # systemd: string, regular expression to use when discovering nodes run by systemd
+    systemd: metrikad-(flow|solana).service
+
+    # docker: string, regular expression to use when discovering nodes run by Docker
+    docker: consensus_1-1
```
Comment (tomasger) [inline configs/agent.yml:126]: Any config changes should also reflect in the ansible folder.

**Review by tomasger — CHANGES_REQUESTED** [review]

File: cmd/agent/main.go [hunk 5]
```diff
@@ -106,18 +108,67 @@ func setupZapLogger() zap.AtomicLevel {
...
-			Regex: regex,
-		}))
+		conf := watch.ContainerWatchConf{
+			Discoverer: dsc,
+			Regex:      confRegex,
+		}
+
+		w, err := watch.NewContainerWatch(conf)
+		if err != nil {
+			zap.S().Fatalw("cannot build container watcher without a discoverer configured", zap.Error(err))
```
[the remark this answers is not in the document]
Comment (iogakos) [reply cmd/agent/main.go]: Addressed in 264b24eec14d.

File: cmd/agent/main.go [hunk 6]
```diff
@@ -160,7 +211,43 @@ func registerWatchers() error {
...
+			scheme := dsc.DetectScheme(ctx)
+
+			switch scheme {
+			case global.NodeDocker:
+				watchersEnabled = append(watchersEnabled, defaultDockerWatchers()...)
+			case global.NodeSystemd:
+				watchersEnabled = append(watchersEnabled, defaultSystemdWatchers()...)
+			default:
+				zap.S().Warnw("node discovery returned no errors but scheme is unknown, retrying", "scheme", scheme)
+				continue
```
[the remark this answers is not in the document]
Comment (iogakos) [reply cmd/agent/main.go]: Addressed in 264b24eec14d.

File: cmd/agent/main.go [hunk 7]
```diff
@@ -160,7 +211,43 @@ func registerWatchers() error {
...
+			}
+			global.BlockchainNode.SetRunScheme(scheme)
+
+			if len(watchersEnabled) > 0 {
+				for _, w := range watchersEnabled {
+					if err := watch.DefaultWatchRegistry.RegisterAndStart(w, subscriptions...); err != nil {
+						zap.S().Errorw("error registering node discovery watchers, retrying in 2s", zap.Error(err))
```
[the remark this answers is not in the document]
Comment (iogakos) [reply cmd/agent/main.go]: Addressed in 264b24eec14d.

File: configs/agent.yml [hunk 8]
```diff
@@ -21,7 +21,7 @@ platform:
   transport_timeout: 10s
 
   # addr: string, network address of the platform for publishing
-  addr: <platform_addr>
+  addr: <platform_addr
```
[the remark this answers is not in the document]
Comment (iogakos) [reply configs/agent.yml]: Addressed in 264b24eec14d.

File: cmd/agent/main.go [hunk 9]
```diff
@@ -58,6 +59,7 @@ var (
 	ctx, pubCtx       context.Context
 	cancel, pubCancel context.CancelFunc
 	promHandler       = promhttp.HandlerFor(prometheus.DefaultGatherer, promhttp.HandlerOpts{EnableOpenMetrics: true})
+	dsc               *utils.NodeDiscoverer
```
[the remark this answers is not in the document]
Comment (iogakos) [reply cmd/agent/main.go]: Addressed in 264b24eec14d.

File: configs/agent.yml [hunk 10]
```diff
@@ -6,7 +6,7 @@ platform:
   enabled: true
 
   # api_key: authentication key for the Metrika platform
-  api_key: <api_key>
+  api_key: <api_key
```
[the remark this answers is not in the document]
Comment (iogakos) [reply configs/agent.yml:9]: Addressed in 264b24eec14d.

File: configs/agent.yml [hunk 11]
```diff
@@ -55,7 +55,7 @@ runtime:
     # level: logging level for runtime logs.
     #
     # Possible values are: info (recommended), warning, debug, error.
-    level: info
+    level: debug
```
[the remark this answers is not in the document]
Comment (iogakos) [reply configs/agent.yml]: Addressed in 6948ed2.

File: configs/agent.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (iogakos) [reply configs/agent.yml]: Addressed in 6948ed2.

File: flow/flow.go [hunk 12]
```diff
@@ -525,3 +494,149 @@ func (d *Flow) Protocol() string {
...
+
+	log.Info("flow node discovery successful")
+
+	return nil
+}
+
+// Reconfigure re-runs the configuration process for all node
+// metadata (i.e. node version).
+func (d *Flow) Reconfigure() error {
+	switch d.runScheme {
```
[the remark this answers is not in the document]
Comment (iogakos) [reply flow/flow.go:604]: Addressed in 6948ed2.

File: flow/flow.go [hunk 13]
```diff
@@ -525,3 +494,149 @@ func (d *Flow) Protocol() string {
...
+	d.container = container
+}
+
+// SetSystemdService sets the node's systemd service unit
+func (d *Flow) SetSystemdService(unit *dbus.UnitStatus) {
+	d.mutex.Lock()
+	defer d.mutex.Unlock()
+
+	d.systemdService = unit
+}
```
[the remark this answers is not in the document]
Comment (iogakos) [reply flow/flow.go:642]: Addressed in 6948ed2.

File: internal/pkg/discover/mock.go [hunk 14]
```diff
@@ -86,10 +88,10 @@ func (m *MockBlockchain) NodeVersion() string {
...
 }
 
-// DiscoverContainer returns the container discovered or an error if any occurs
-func (m *MockBlockchain) DiscoverContainer() (*types.Container, error) {
-	return &types.Container{Names: []string{"/flow-private-network_consensus_3_1"}}, nil
-}
+// // DiscoverContainer returns the container discovered or an error if any occurs
+// func (m *MockBlockchain) DiscoverContainer() (*types.Container, error) {
+// 	return &types.Container{Names: []string{"/flow-private-network_consensus_3_1"}}, nil
+// }
```
[the remark this answers is not in the document]
Comment (iogakos) [reply internal/pkg/discover/mock.go]: Addressed in 6948ed2.

File: internal/pkg/discover/utils/node.go [hunk 15]
```diff
@@ -0,0 +1,254 @@
...
+	ErrNodeServiceNotFound = errors.New("no node run by systemd detected")
+
+	// ErrNodeDiscovererConfig node discoverer configuration error
+	ErrNodeDiscovererConfig = errors.New("at least one regular expression is required (i.e. containerRegex)")
+
+	// ErrNodeDiscoveryCancelled node discovery cancelled error
+	ErrNodeDiscoveryCancelled = errors.New("node discovery cancelled")
+
+	// SupportedSchemes list of supported running schemes
+	SupportedSchemes = []global.NodeRunScheme{global.NodeDocker, global.NodeSystemd}
```
[the remark this answers is not in the document]
Comment (iogakos) [reply internal/pkg/discover/utils/node.go]: Addressed in 6948ed2.

File: internal/pkg/watch/watch.go (hunk 1 above)
Comment (iogakos) [reply internal/pkg/watch/watch.go]: Addressed in 264b24eec14d.

File: internal/pkg/watch/journal_logs.go (hunk 2 above)
Comment (iogakos) [reply internal/pkg/watch/journal_logs.go]: Addressed in 264b24eec14d.

File: protobind/node.go.template (hunk 3 above)
Comment (iogakos) [reply protobind/node.go.template]: Addressed in 264b24eec14d.

File: configs/agent.yml (hunk 4 above)
Comment (iogakos) [reply configs/agent.yml]: Addressed in ddaa36a.

**Review by tomasger — APPROVED** [review]
LGTM!

**Review by tomasger — DISMISSED** [review]

**Review by tomasger — APPROVED** [review]

**Pull request merged**